app/command_invite.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"strings"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/mattermost-server/model"
+	goi18n "github.com/nicksnyder/go-i18n/i18n"
+)
+
+type InviteProvider struct {
+}
+
+const (
+	CMD_INVITE = "invite"
+)
+
+func init() {
+	RegisterCommandProvider(&InviteProvider{})
+}
+
+func (me *InviteProvider) GetTrigger() string {
+	return CMD_INVITE
+}
+
+func (me *InviteProvider) GetCommand(a *App, T goi18n.TranslateFunc) *model.Command {
+	return &model.Command{
+		Trigger:          CMD_INVITE,
+		AutoComplete:     true,
+		AutoCompleteDesc: T("api.command_invite.desc"),
+		AutoCompleteHint: T("api.command_invite.hint"),
+		DisplayName:      T("api.command_invite.name"),
+	}
+}
+
+func (me *InviteProvider) DoCommand(a *App, args *model.CommandArgs, message string) *model.CommandResponse {
+	if message == "" {
+		return &model.CommandResponse{Text: args.T("api.command_invite.missing_message.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	l4g.Debug(message)
+
+	splitMessage := strings.SplitN(message, " ", 2)
+	targetUsername := splitMessage[0]
+	targetUsername = strings.TrimPrefix(targetUsername, "@")
+
+	var userProfile *model.User
+	if result := <-a.Srv.Store.User().GetByUsername(targetUsername); result.Err != nil {
+		l4g.Error(result.Err.Error())
+		return &model.CommandResponse{Text: args.T("api.command_invite.missing_user.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	} else {
+		userProfile = result.Data.(*model.User)
+	}
+
+	var channelToJoin *model.Channel
+	var err *model.AppError
+	// User set a channel to add the invited user
+	if len(splitMessage) > 1 && splitMessage[1] != "" {
+		targetChannelName := strings.TrimPrefix(strings.TrimSpace(splitMessage[1]), "~")
+
+		if channelToJoin, err = a.GetChannelByName(targetChannelName, args.TeamId); err != nil {
+			return &model.CommandResponse{Text: args.T("api.command_invite.channel.error", map[string]interface{}{"Channel": targetChannelName}), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+		}
+	} else {
+		channelToJoin, err = a.GetChannel(args.ChannelId)
+		if err != nil {
+			return &model.CommandResponse{Text: args.T("api.command_invite.channel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+		}
+	}
+
+	// Check if is a Direct Channel
+	if channelToJoin.Type == model.CHANNEL_DIRECT || channelToJoin.Type == model.CHANNEL_GROUP {
+		return &model.CommandResponse{Text: args.T("api.command_invite.directchannel.app_error"), ResponseType: model.COMMAND_RESPONSE_TYPE_EPHEMERAL}
+	}
+
+	// Check if user is already in the channel
+	userChannels, err := a.GetChannelsForUser(args.TeamId, userProfile.Id)
```

Comment:
Why not check directly for membership existence? instead of loop over all the user channels something like `a.GetChannelMember(channelToJoin.Id, userProfile.Id); if err == nil { return error}`.

Reply:
good tip. I was looking for something, but did not think about this one 😄  thanks so much! will update that

Reply:
Done. PTAL